plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryTypeManager.java

```diff
@@ -355,31 +369,98 @@ private Optional<ColumnMapping> convertToTrinoType(Field field)
+                return Optional.of(new ColumnMapping(TIMESTAMP_MICROS, true));
             case TIME:
-                return Optional.of(new ColumnMapping(TimeType.TIME_MICROS, true));
+                return Optional.of(new ColumnMapping(TIME_MICROS, true));
             case TIMESTAMP:
-                return Optional.of(new ColumnMapping(TimestampWithTimeZoneType.TIMESTAMP_TZ_MICROS, true));
+                return Optional.of(new ColumnMapping(TIMESTAMP_TZ_MICROS, true));
             case GEOGRAPHY:
-                return Optional.of(new ColumnMapping(VarcharType.VARCHAR, false));
+                return Optional.of(new ColumnMapping(VARCHAR, false));
             case JSON:
                 return Optional.of(new ColumnMapping(jsonType, false));
             case STRUCT:
-                // create the row
-                FieldList subTypes = field.getSubFields();
-                checkArgument(!subTypes.isEmpty(), "a record or struct must have sub-fields");
-                List<RowType.Field> fields = subTypes.stream().map(subField -> toRawTypeField(subField.getName(), subField)).collect(toList());
-                RowType rowType = RowType.from(fields);
-                return Optional.of(new ColumnMapping(rowType, false));
+                return Optional.of(new ColumnMapping(createRowType(field), false));
             default:
                 return Optional.empty();
         }
     }
 
+    private RowType createRowType(Field field)
+    {
+        FieldList subTypes = field.getSubFields();
+        checkArgument(!subTypes.isEmpty(), "a record or struct must have sub-fields");
+        List<RowType.Field> fields = subTypes.stream()
+                .map(subField -> toRawTypeField(subField.getName(), subField))
+                .collect(toImmutableList());
+        return RowType.from(fields);
+    }
+
+    public List<ColumnMetadata> convertToTrinoType(List<String> names, List<String> types)
+    {
+        return convertToTrinoType(names, types, Optional::empty);
+    }
+
+    public List<ColumnMetadata> convertToTrinoType(List<String> names, List<String> types, Supplier<Optional<TableInfo>> tableSupplier)
+    {
+        checkArgument(names.size() == types.size(), "Mismatched column names and types");
+
+        ImmutableList.Builder<ColumnMetadata> columns = ImmutableList.builder();
+        for (int i = 0; i < names.size(); i++) {
+            String name = names.get(i);
+            TypeSignature typeSignature;
+            try {
+                TypeInfo typeInfo = getTypeInfoFromTypeString(types.get(i));
+                typeSignature = toTypeSignature(typeInfo);
+            }
+            catch (UnsupportedTypeException e) {
+                Optional<TableInfo> table = tableSupplier.get();
+                if (!e.getTypeName().equals(STRUCT) || table.isEmpty() || table.get().getDefinition().getSchema() == null) {
+                    // ignore unsupported types
+                    continue;
```

Comment:
Why do we want to continue iteration even when table is empty?

Reply:
We only need the table to map STRUCT types. We should continue iteration to convert other types.

Reply:
I suppose `table.isEmpty()` means the table doesn't exist. No need to return columns in my opinion.

Reply:
It doesn't mean it doesn't exit, just it wasn't provided.

Reply:
No, the `tableSupplier` is called with `getTable` method:
https://github.com/trinodb/trino/blob/569b045f3bb228703f93a4784411274df5f597d7/plugin/trino-bigquery/src/main/java/io/trino/plugin/bigquery/BigQueryClient.java#L208-L217